### PR TITLE
Localize the text in the common target

### PR DESCRIPTION
**Pull request by JaynieBai**
Fixes [#10171 ](https://github.com/dotnet/msbuild/issues/10171)

### Context
MsBuild produces unlocalized messages for some specific validation cases that happen in scope of target.
The example of the task usage: https://github.com/YuliiaKovalova/msbuild/blob/be21253d85f7766356880d376e26aaa69c34c4cd/src/Tasks/Microsoft.Common.CurrentVersion.targets#L862

### Changes Made
Refactor the existing messages with the recently added task allows to produce localized messages:
https://github.com/dotnet/msbuild/blob/79dff86b18613cfe3510b719ac28e8a8c3e7f96c/src/Tasks/MSBuildInternalMessage.cs , produced by with hardcoded text.
### Testing
Add test cases for each new resource, except for the resource CommonSdk.PropertyWithTrailingSlash, which is tested manually as followed picture.
image

### CODE REVIEW

**Review by JanKrivanek — APPROVED** [review]
Looks good.

Though - I was not controllig in detail 1:1 equality of the literals in .targets with the new in resx files - I suppose they were copied over (with just applying format parameters)

**Review by Nirmal4G — CHANGES_REQUESTED** [review]
Why not use `CommonSdk` instead of `CommonTarget`? It could be useful if you decide to do #1686?

Also, Message Code (like `MSB3539`) should be separately specified to surface them properly in VS IDE's Messages and VS Code's Problems panels.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 1]
```diff
@@ -834,23 +834,20 @@ Copyright (C) Microsoft Corporation. All rights reserved.
   <Target
       Name="_CheckForInvalidConfigurationAndPlatform"
       BeforeTargets="$(BuildDependsOn);Build;$(RebuildDependsOn);Rebuild;$(CleanDependsOn);Clean">
-
```
Comment (Nirmal4G) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:837]: nit: undo the line deletion!

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 2]
```diff
@@ -834,23 +834,20 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+      <_InvalidConfigurationMessageResourceName Condition=" '$(BuildingInsideVisualStudio)' == 'true' ">CommonSdk.InvalidConfigurationTextWhenBuildingInsideVisualStudio</_InvalidConfigurationMessageResourceName>
+      <_InvalidConfigurationMessageResourceName Condition=" '$(BuildingInsideVisualStudio)' != 'true' ">CommonSdk.InvalidConfigurationTextWhenBuildingOutsideVisualStudio</_InvalidConfigurationMessageResourceName>
     </PropertyGroup>
 
-    <Error Condition=" '$(_InvalidConfigurationError)' == 'true' " Text="$(_InvalidConfigurationMessageText)"/>
-    <Warning Condition=" '$(_InvalidConfigurationWarning)' == 'true' " Text="$(_InvalidConfigurationMessageText)"/>
+    <MSBuildInternalMessage ResourceName="$(_InvalidConfigurationMessageResourceName)" Severity="$(_InvalidConfigurationMessageSeverity)" FormatArguments="$(MSBuildProjectFile);$(_OriginalConfiguration);$(_OriginalPlatform)"/>
```
Comment (Nirmal4G) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:842]: Sorry about my mistake.

```suggestion
    <MSBuildInternalMessage
      ResourceName="$(_InvalidConfigurationMessageResourceName)"
      Severity="$(_InvalidConfigurationMessageSeverity)"
      FormatArguments="$(MSBuildProjectFile);$(_OriginalConfiguration);$(_OriginalPlatform)"
      Condition="'$(_InvalidConfigurationMessageSeverity)' != ''" />
```

> [!IMPORTANT]
>
> For the condition, we could have used the same output path check here also
> but we must not forget to update both when it changes in the future.
> That's why I have used Severity instead.

**Review by Nirmal4G — CHANGES_REQUESTED** [review]

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 3]
```diff
@@ -2171,16 +2179,17 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-        Condition="'@(ProjectReferenceWithConfiguration)' != '' and '@(_MSBuildProjectReferenceNonexistent)' != '' and '$(ErrorOnMissingProjectReference)' == 'True'"/>
+    <!-- Issue a warning for each non-existent project. -->
+    <PropertyGroup>
+      <_NonExistentProjectReferenceSecuity  Condition="'@(ProjectReferenceWithConfiguration)' != '' and '@(_MSBuildProjectReferenceNonexistent)' != '' and '$(ErrorOnMissingProjectReference)' != 'True'">Warning</_NonExistentProjectReferenceSecuity>
+      <_NonExistentProjectReferenceSecuity  Condition="'@(ProjectReferenceWithConfiguration)' != '' and '@(_MSBuildProjectReferenceNonexistent)' != '' and '$(ErrorOnMissingProjectReference)' == 'True'">Error</_NonExistentProjectReferenceSecuity>
+    </PropertyGroup>
+    <MSBuildInternalMessage
+      Condition="'$(_NonExistentProjectReferenceSecuity)' != ''"
+      ResourceName="CommonSdk.NonExistentProjectReference"
+      Severity="$(_NonExistentProjectReferenceSecuity)"
```
Comment (Nirmal4G) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:2190]:
```suggestion
      Severity="$(_NonExistentProjectReferenceSeverity)"
```

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 2 above)
Comment (JaynieBai) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets]: Thanks a lot @Nirmal4G

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 4]
```diff
@@ -835,22 +835,18 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+    <MSBuildInternalMessage Condition=" '$(_InvalidConfigurationError)' == 'true' and '$(BuildingInsideVisualStudio)' != 'true'" ResourceName="CommonTarget.InvalidConfigurationTextWithoutBuildingInsideVisualStudio" Severity="Error" FormatArguments="$(MSBuildProjectFile);$(_OriginalConfiguration);$(_OriginalPlatform)"/>
+    <MSBuildInternalMessage Condition=" '$(_InvalidConfigurationWarning)' == 'true' and '$(BuildingInsideVisualStudio)' == 'true'" ResourceName="CommonTarget.InvalidConfigurationTextWithBuildingInsideVisualStudio" Severity="Warning" FormatArguments="$(MSBuildProjectFile);$(_OriginalConfiguration);$(_OriginalPlatform)"/>
+    <MSBuildInternalMessage Condition=" '$(_InvalidConfigurationWarning)' == 'true' and '$(BuildingInsideVisualStudio)' != 'true'" ResourceName="CommonTarget.InvalidConfigurationTextWithoutBuildingInsideVisualStudio" Severity="Warning" FormatArguments="$(MSBuildProjectFile);$(_OriginalConfiguration);$(_OriginalPlatform)"/>
```
[the remark this answers is not in the document]
Comment (JaynieBai) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets]: Updated

File: src/Tasks.UnitTests/MSBuildInternalMessage_Tests.cs [hunk 5]
```diff
@@ -59,5 +60,233 @@ public void E2EScenarioTests(bool prefer32, bool isPlatformAnyCpu, string expect
                 }
             }
         }
+
+        [Theory]
+        [InlineData(true, "CommonSdk.BaseIntermediateOutputPathMismatchWarning")]
+        [InlineData(false, "CommonSdk.MSBuildProjectExtensionsPathModifiedAfterUse")]
+
+        public void BaseIntermediateOutputPathMisMatchWarning(bool IsInitialMSBuildProjectExtensionsPathEmpty, string expectedResourceName)
```
Comment (Nirmal4G) [inline src/Tasks.UnitTests/MSBuildInternalMessage_Tests.cs:68]:
```suggestion
        public void BaseIntermediateOutputPathMisMatchWarning(bool isInitialMSBuildProjectExtensionsPathEmpty, string expectedResourceName)
```

File: src/Tasks.UnitTests/MSBuildInternalMessage_Tests.cs [hunk 6]
```diff
@@ -59,5 +60,233 @@ public void E2EScenarioTests(bool prefer32, bool isPlatformAnyCpu, string expect
...
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                var outputPath = env.CreateFolder().Path;
+                string projectContent = $"""
+                <Project>
+                    <Import Project="$(MSBuildBinPath)\Microsoft.Common.props" />
+
+                    <PropertyGroup>
+                        <EnableBaseIntermediateOutputPathMismatchWarning>true</EnableBaseIntermediateOutputPathMismatchWarning>
+                        <_InitialMSBuildProjectExtensionsPath>{(IsInitialMSBuildProjectExtensionsPathEmpty ? "" : "obj")}</_InitialMSBuildProjectExtensionsPath>
```
Comment (Nirmal4G) [inline src/Tasks.UnitTests/MSBuildInternalMessage_Tests.cs:79]:
```suggestion
                        <_InitialMSBuildProjectExtensionsPath>{(isInitialMSBuildProjectExtensionsPathEmpty ? "" : "obj")}</_InitialMSBuildProjectExtensionsPath>
```

File: src/Tasks.UnitTests/MSBuildInternalMessage_Tests.cs [hunk 7]
```diff
@@ -59,5 +60,233 @@ public void E2EScenarioTests(bool prefer32, bool isPlatformAnyCpu, string expect
...
+                """;
+
+                var projectFile = env.CreateFile(env.CreateFolder(), "test.csproj", projectContent).Path;
+                Project project = ObjectModelHelpers.LoadProjectFileInTempProjectDirectory(projectFile, touchProject: false);
+
+                string expectedBuildMessage = ResourceUtilities.FormatResourceStringStripCodeAndKeyword(expectedResourceName);
+                MockLogger logger = new MockLogger(_testOutput);
+
+                project.Build(logger);
+                if (!IsInitialMSBuildProjectExtensionsPathEmpty)
```
Comment (Nirmal4G) [inline src/Tasks.UnitTests/MSBuildInternalMessage_Tests.cs:95]:
```suggestion
                if (!isInitialMSBuildProjectExtensionsPathEmpty)
```

**Review by Nirmal4G — CHANGES_REQUESTED** [review]
Code Style nit: Method parameter should start with lower case.

Other that that, it's looking good.

File: src/Tasks.UnitTests/MSBuildInternalMessage_Tests.cs (hunk 6 above)
Comment (JaynieBai) [reply src/Tasks.UnitTests/MSBuildInternalMessage_Tests.cs]: Sorry. missed that. Updated

**Comment by JaynieBai** [comment]
> > Message Code (like MSB3539) should be separately specified to ensure searchability in these files.
> 
> We only localize the text, so, it's not strictly required to put `Code` in. But I get the reason why. I can also argue that, when debugging an obscure build failure, you might need to locate the line in the targets where it spits out the message and start from there. Since, for almost all messages, the text log does not provide line number and file of the message task unless a high verbosity level is provided. So, you cannot locate the message task.
> 
> With `Code` specified in the targets, we can just do a text search on a preprocessed file, and we can go to that exact line. This saves a ton of time combing through binlog or imported files.
> 
> And if you want to indicate the `Code` mapping in RESX, why not add it in the comments? There's a reason that previous text only message tasks had a separate `Code` parameter instead of just putting them in the `Text` parameter.

Thanks for your explanation. I added that code in the comment.

**Comment by Nirmal4G** [comment]
> I added that code in the comment.

This works just as fine, but it requires anyone who adds a message task to consistently place the `Code` in the comments, which is improbable. IMO, separating the `Code` from the resource string remains most effective.

Also, `Code` in the resource string seems to be optional even if the Severity is either `Warning` or `Error`. Since this is an internal task to MSBuild, can't we make the `Code` mandatory on those conditions? This would prevent some typos in the future and also force us to allocate `Code` for every warning or error.

**Review by Nirmal4G — APPROVED** [review]

**Comment by JanKrivanek** [comment]
@Nirmal4G thonks for nice suggestions - tracked as https://github.com/dotnet/msbuild/issues/11087 - as the change would require adjusting code violating this currently and is out of scope of this PR

**Review by YuliiaKovalova — APPROVED** [review]

**Comment by Nirmal4G** [comment]
@JanKrivanek No problem. As you said, these are **just** suggestions. And I ~~thonk~~ thank you (sorry, couldn't resist 😜) for considering it.

**Comment by JanKrivanek** [comment]
:laughing: appology - quick typing. I promise I gave more time into considering the suggestion than into to typing that response! 😉 (it's tracked after all)

**Pull request merged**